lib/parser/source/buffer.rb

```diff
@@ -303,27 +296,41 @@ def source_range
       # @return [Integer]
       #
       def last_line
-        line_begins.size + @first_line - 1
+        line_begins.size + @first_line - 2
       end
 
       private
 
+      # @returns [0, line_begin_of_line_1, ..., source.size + 1]
       def line_begins
-        unless @line_begins
-          @line_begins, index = [ [ 0, 0 ] ], 0
-
+        @line_begins ||= begin
```

Comment:
looks like begin body should be changed to not mutate `@line_begin`. Also, it implicitly relies on return value of `Array#<<`

Reply:
👍 tweaked